File: solutions/cpp/0373.cpp
class Solution {
 public:
  vector<vector<int>> kSmallestPairs(vector<int>& nums1, vector<int>& nums2,
                                     int k) {
    if (nums2.empty()) return {};

    vector<vector<int>> ans;

    auto compare = [&](const pair<int, int>& a, const pair<int, int>& b) {
      return nums1[a.first] + nums2[a.second] >
             nums1[b.first] + nums2[b.second];
    };
    priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(compare)>
        pq(compare);

    for (int i = 0; i < nums1.size() && i < k; ++i)
      pq.push({i, 0});

    while (!pq.empty() && ans.size() < k) {
      const auto [index1, index2] = pq.top(); pq.pop();
      ans.push_back({nums1[index1], nums2[index2]});
      if (index2 + 1 < nums2.size())
        pq.push({index1, index2 + 1});
    }

    return ans;
  }
};

File: solutions/java/0373.java
class Solution {
  public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
    if (nums1.length == 0 || nums2.length == 0 || k == 0)
      return new ArrayList<>();

    List<List<Integer>> ans = new ArrayList<>();
    PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> (nums1[a[0]] + nums2[a[1]]) - (nums1[b[0]] + nums2[b[1]]));

    for (int i = 0; i < nums1.length && i < k; ++i)
      pq.offer(new int[] { i, 0 });

    while (k-- > 0 && !pq.isEmpty()) {
      int[] index = pq.poll();
      ans.add(Arrays.asList(nums1[index[0]], nums2[index[1]]));
      if (index[1] < nums2.length - 1)
        pq.offer(new int[] { index[0], index[1] + 1 });
    }

    return ans;
  }
}